odometry.c

```diff
@@ -0,0 +1,190 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <math.h>
+
+#include "timestamp/timestamp.h"
+#include "odometry.h"
+
+#define MINIMUM_DELTA_T 1e-7 // To avoid dividing by zero in derivative
+
+
+void base_init(
+		odometry_differential_base_t *robot,
+		const struct robot_base_pose_2d_s init_pose,
+		const float right_wheel_radius,
+		const float left_wheel_radius,
+		const float wheelbase,
+		const timestamp_t time_now)
+{
+	robot->pose.x = init_pose.x;
+	robot->pose.y = init_pose.y;
+	robot->pose.theta = init_pose.theta;
+
+	robot->velocity.x = 0.0f;
+	robot->velocity.y = 0.0f;
+	robot->velocity.omega = 0.0f;
+
+	wheel_init(&robot->right_wheel, right_wheel_radius);
+	wheel_init(&robot->left_wheel, left_wheel_radius);
+
+	robot->wheelbase = wheelbase;
+	robot->time_last_estim = time_now;
+}
+
+void base_update(
+		odometry_differential_base_t *robot,
+		const odometry_encoder_sample_t right_wheel_sample,
+		const odometry_encoder_sample_t left_wheel_sample)
+{
+	wheel_update(&robot->right_wheel, right_wheel_sample);
+	wheel_update(&robot->left_wheel, left_wheel_sample);
+
+	// Choose wheel with most recent timestamp as reference
+	timestamp_t time_now;
+	if (0.0f >= timestamp_duration_s(right_wheel_sample.timestamp,
+									 left_wheel_sample.timestamp)) {
+		time_now = right_wheel_sample.timestamp;
+	} else {
+		time_now = left_wheel_sample.timestamp;
+	}
+
+	// Wheel prediction
+	float delta_right_wheel = wheel_get_delta_meter(&robot->right_wheel, time_now);
+	float delta_left_wheel = wheel_get_delta_meter(&robot->left_wheel, time_now);
+
+	// Base prediction
+	float dt = timestamp_duration_s(robot->time_last_estim, time_now);
+	float delta_fwd = 0.5f * (delta_right_wheel + delta_left_wheel);
+	float delta_rot = (delta_right_wheel - delta_left_wheel) / robot->wheelbase;
+
+	if (dt >= MINIMUM_DELTA_T) {
+		robot->velocity.x = delta_fwd
+							 * cos(robot->pose.theta + 0.5f * delta_rot) / dt;
+		robot->velocity.y = delta_fwd
+							 * sin(robot->pose.theta + 0.5f * delta_rot) / dt;
+		robot->velocity.omega = delta_rot / dt;
+	}
+
+	robot->pose.x += delta_fwd * cos(robot->pose.theta + 0.5f * delta_rot);
+	robot->pose.y += delta_fwd * sin(robot->pose.theta + 0.5f * delta_rot);
```

Comment:
Actually you might want to use the `cost` function to avoid type promotion to double.